src/Lecture2.hs

```diff
@@ -158,14 +174,97 @@ You're free to define any helper functions.
 -}
 
 -- some help in the beginning ;)
-data Knight = Knight
-    { knightHealth    :: Int
-    , knightAttack    :: Int
-    , knightEndurance :: Int
-    }
-
-dragonFight = error "TODO"
 
+data Knight = Knight
+  { knightHealth :: Int
+  , knightAttack :: Int
+  , knightEndurance :: Int
+  }
+
+data Chest a = Chest
+  { gold :: Int
+  , treasure :: Maybe a
+  }
+
+data DragonType = Black | Green | Red
+data Dragon a = Dragon
+  { dragonType :: DragonType
+  , dragonFirePower :: Int
+  , dragonHealth :: Int
+  , swallowedChest :: Chest a
+  }
```

Comment:
The solution for making illegal states unrepresentable without using GADTs is to use a simple record type:

```haskell
data DragonProperties = DragonProperties { ... } -- all common fields of different dragon types

data Dragon a
    = Black DragonProperties (Chest a)
    | Red DragonProperties (Chest a)
    | Green DragonProperties Int  -- storing only Gold here
```

Reply:
Ah I feel so dumb for not thinking about this sooner! 😂